modules/avatar/identicon/polygon.go

```diff
@@ -0,0 +1,71 @@
+// Copyright 2021 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+// Copied and modified from https://github.com/issue9/identicon/ (MIT License)
+
+package identicon
+
+var (
+	// cos(0),cos(90),cos(180),cos(270)
+	cos = []int{1, 0, -1, 0}
+
+	// sin(0),sin(90),sin(180),sin(270)
+	sin = []int{0, 1, 0, -1}
+)
+
+// rotate the points by center point (x,y)
+// angle: [0,1,2,3] means [0，90，180，270] degree
+func rotate(points []int, x, y int, angle int) {
+	if angle < 0 || angle > 3 {
+		panic("rotate: angle must be 0/1/2/3")
+	}
+
+	for i := 0; i < len(points); i += 2 {
+		px, py := points[i]-x, points[i+1]-y
+		points[i] = px*cos[angle] - py*sin[angle] + x
+		points[i+1] = px*sin[angle] + py*cos[angle] + y
+	}
+}
+
+// check whether the point is inside the point is in the polygon (points)
+// the first and the last point must be the same
+func pointInPolygon(x, y int, polygonPoints []int) bool {
+	if len(polygonPoints) < 8 { // a valid polygon must have more than 2 points
+		return false
+	}
+
+	// split the plane into two by the check point horizontally:
```

Comment:
Just to have it clear for anyone going to read this code ever, this code seems to be based on the [nonzero winding rule](https://en.wikipedia.org/wiki/Nonzero-rule). If so, we should note it somewhere such that if someone doesn't make sense of this code that they can lookup the principle behind this algorithm